ENT-4192 Identity service refactor for confidential-identities and accounts

roger-that-dev opened the pull request:
Throw stones as necessary...

1. Moved `externalIdForPublicKey` from `KeyManagementService` to `IdentityService` as request by Matthew as the semantics didn't really work so well with it sitting in the KMS.
2. Added a new API to `IdentityService` called `registerKeyToExternalId` which allows CorDapp developers to map `PublicKey`s to `UUID`s
3. Added an API to return all the keys associated with a specified UUID.

Commit:
Removed IdentityServiceInternal as it is no longer used.

Commit:
Removed externalIdForPublicKey API from KMS and added it to IdentityService.
Added a registerKeyToExternalId API on IdentityService.

Commit:
Fix remaining compile errors.

Commit:
Removed "registerKeyToParty" and in its place added a new registerKey method which takes a PublicKey, Party and optionally a UUID.
Added a cache to the "PersistentIdentityService" to store other node's public keys.
Added the cache and new hibernate entity to all teh places where one needs to add them.
New keys created by teh node now automatically get associated entries in the KEY -> PARTY map and optionally the KEy -> EXT ID map.
Added a test.

Commit:
Removed old comments and TODOs.

Commit:
Fixed broken test.
Added comments/explanations for what's going on in IdentityService.
Updated kdocs.

Commit:
First try at Implementing publicKeysForExternalId.

Commit:
Fixed broken test.

mnesbit reviewed (dismissed):
Makes sense to me

kasiastreich commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -68,8 +70,6 @@ interface IdentityService {
      * @param owningKey The [PublicKey] to determine well known identity for.
      * @return the party and certificate, or null if unknown.
      */
-    @Deprecated("This method has been deprecated in favour of using a new way to generate and use confidential identities. See the new " +
```
So do we deprecate it or not?

roger-that-dev replied:
We can't really deprecate this because it is still used by a few other things.

kasiastreich commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -150,15 +150,32 @@ interface IdentityService {
+     * @throws IllegalArgumentException if the public publicKey is already registered to a party that does not match the supplied party
      */
     @Throws(IllegalArgumentException::class)
-    fun registerKeyToParty(key: PublicKey, party: Party)
```
Isn't it API break?

roger-that-dev replied:
It usually would be but this is unreleased code - so we can still move stuff around :)

kasiastreich commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -150,15 +150,32 @@ interface IdentityService {
-     * exists for the supplied [PublicKey] but the associated [Party] does not match the one supplied to the method then an exception will
-     * be thrown.
+     * exists for the supplied [PublicKey] but the associated [Party] does not match the one supplied to the method then a warning will be
+     * logged and the operation will not be carried out as a key can only ever be registered to one [Party].
```
But you still throw an exception in this case?

roger-that-dev replied:
No, not anymore. I don't think it's an exceptional case. What do you think ?

roger-that-dev replied:
Added it back - thanks

kasiastreich commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -150,15 +150,32 @@ interface IdentityService {
      */
     @Throws(IllegalArgumentException::class)
-    fun registerKeyToParty(key: PublicKey, party: Party)
+    fun registerKey(publicKey: PublicKey, party: Party, externalId: UUID? = null)
```
Shouldn't we add `@JvmOverloads` if you introduce API method with defaults?

roger-that-dev replied:
ok

kasiastreich commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -150,15 +150,32 @@ interface IdentityService {
+    @Suspendable
+    fun externalIdForPublicKey(publicKey: PublicKey): UUID?
+
+    fun publicKeysForExternalId(externalId: UUID): Iterable<PublicKey>
```
add kdocs for this method too

roger-that-dev replied:
It's not exposed via the API but will add docs

kasiastreich commented on node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt:
```diff
@@ -108,10 +111,19 @@ class InMemoryIdentityService(identities: List<PartyAndCertificate> = emptyList(
-            keyToName.putIfAbsent(key, party.name)
-        }
-        throw IllegalArgumentException("An entry for the public key: $key already exists.")
+    override fun registerKey(publicKey: PublicKey, party: Party, externalId: UUID?) {
```
Don't you have to implement it?

roger-that-dev replied:
True, I was just being lazy :P

roger-that-dev replied:
Implemented now.

kasiastreich commented on node/src/main/kotlin/net/corda/node/services/identity/PersistentIdentityService.kt:
```diff
@@ -94,6 +102,23 @@ class PersistentIdentityService(cacheFactory: NamedCacheFactory) : SingletonSeri
+                                Crypto.decodePublicKey(it.publicKey)
+                        )
+                    },
+                    toPersistentEntity = { key: String, value: PublicKey ->
```
Shouldn't hash be derived from the key?

roger-that-dev replied:
Yeah, it is. I guess you are saying that the constructor for this can be simplified by just providing a public key instead of a publickey and a string (which is the hash).

kasiastreich commented on node/src/main/kotlin/net/corda/node/services/identity/PersistentIdentityService.kt:
```diff
@@ -289,18 +338,70 @@ class PersistentIdentityService(cacheFactory: NamedCacheFactory) : SingletonSeri
-            val existingEntryForKey = keyToName[key.toStringShort()]
+            val publicKeyHash = publicKey.toStringShort()
+            // EVERY key should by mapped to a Party in the "keyToName" table. Therefore if there is already a record in that table for the
+            // specified key then it's either our key which has been stored prior or another node's key which we have preciously mapped.
```
Preciously or previously :joy:

roger-that-dev replied:
oops thanks

kasiastreich commented on node/src/main/kotlin/net/corda/node/services/identity/PersistentIdentityService.kt:
```diff
@@ -289,18 +338,70 @@ class PersistentIdentityService(cacheFactory: NamedCacheFactory) : SingletonSeri
         return database.transaction {
-            val existingEntryForKey = keyToName[key.toStringShort()]
+            val publicKeyHash = publicKey.toStringShort()
+            // EVERY key should by mapped to a Party in the "keyToName" table. Therefore if there is already a record in that table for the
```
typo: should be

roger-that-dev replied:
noted

kasiastreich commented on node/src/main/kotlin/net/corda/node/services/identity/PersistentIdentityService.kt:
```diff
@@ -289,18 +338,70 @@ class PersistentIdentityService(cacheFactory: NamedCacheFactory) : SingletonSeri
-                if (party.name != keyToName[key.toStringShort()]) {
-                    throw IllegalArgumentException("The public key ${key.hash} is already assigned to a different party than the supplied .")
+                log.info("An existing entry for $publicKeyHash already exists.")
+                if (party.name != keyToName[publicKeyHash]) {
```
!= existingEntryForKey

kasiastreich commented on node/src/main/kotlin/net/corda/node/services/identity/PersistentIdentityService.kt:
```diff
@@ -289,18 +338,70 @@ class PersistentIdentityService(cacheFactory: NamedCacheFactory) : SingletonSeri
+        return _pkToIdCache[publicKey]?.uuid
+    }
+
+    private fun publicKeysForExternalId(externalId: UUID, table: Class<*>): List<PublicKey> {
```
Could you comment why we do that? It's not clear that we have two tables one for our keys second for the other parties.

roger-that-dev replied:
Yeah I can add a comment. Basically, "our keys" are kept separate from "not our keys" due to the semantics of the KMS being for "our keys" only.

szymonsztuka commented on node/src/main/kotlin/net/corda/node/services/identity/PersistentIdentityService.kt:
```diff
@@ -131,6 +156,18 @@ class PersistentIdentityService(cacheFactory: NamedCacheFactory) : SingletonSeri
+            @Column(name = PK_HASH_COLUMN_NAME, length = MAX_HASH_HEX_SIZE, nullable = false)
+            var publicKeyHash: String = "",
+
+            @Lob
```
With @lob mapping this will create OID column type on PostgreSQL (this is our defult way) however maybe better would be to use BYTE type - there can be a performance gain for selects and it's easier to write raw JDBC if needed.
To change column to be BYTEA in PostgreSQL you would need to add `@Type(type = "corda-blob")` instead of `@Lob` and specific branch in Liquibase XML for Postgres like in file 
`/node/src/main/resources/migration/node-core.changelog-postgres-blob.xml`

roger-that-dev replied:
Thanks @szymonsztuka !

kasiastreich reviewed (commented):
minor comments

roastario commented:
build

roger3cev commented:
Still to do:
1. Implement mock identity service methods.
2. Double check migration is set up as it should be.
3. Add a test for the migration.

Commit:
Added migration.
Amended existing persistent identity service migration to handle new migration.
Addressed some review comments.

Commit:
Fixed broken test - whoops!

willhr3 reviewed (dismissed):
Looks good to me - cheers!

kasiastreich commented on node/src/main/kotlin/net/corda/node/migration/PersistentIdentityMigrationNewTable.kt:
```diff
@@ -41,42 +28,55 @@ class PersistentIdentityMigrationNewTable : CordaMigration() {
+            it.setString(2, name.toString())
             it.executeUpdate()
         }
+        val us = identityService.ourNames.first()
```
are we sure that the first one there will be correct?

roger-that-dev replied:
Someone told me it was!

kasiastreich replied:
it's set, I would check for inclusion

roger-that-dev replied:
OK

kasiastreich commented on node/src/main/kotlin/net/corda/node/migration/PersistentIdentityMigrationNewTable.kt:
```diff
@@ -41,42 +28,55 @@ class PersistentIdentityMigrationNewTable : CordaMigration() {
             it.executeUpdate()
         }
+        val us = identityService.ourNames.first()
+        if (name != us) {
```
or should it be a check `name !in ourNames`?

willhr3 replied:
Yeah I think this would work better _just in case_ theres some DB weirdness and `us` isn't `ourNames.first()`

roger-that-dev replied:
Fixed

kasiastreich commented on node/src/test/kotlin/net/corda/node/services/vault/ExternalIdMappingTest.kt:
```diff
@@ -140,11 +141,72 @@ class ExternalIdMappingTest {
     }
 
+    @Test
+    fun `roger uber keys test`() {
```
fun fun fun :joy:

roger-that-dev replied:
🎉

Commit:
Implemented mock identity service methods.

Commit:
Merge branch 'release/os/4.3' into rog-external-id-refactor

Commit:
Added back exception when remapping a key to a different party.

Commit:
Fixed compile errors.
Fixed broken tests.

willhr3 commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -153,15 +153,32 @@ interface IdentityService {
+     * @param party The party that the supplied public publicKey will be registered to
+     * @param externalId The [UUID] that the supplied public key can be optionally registered to
+     * @throws IllegalArgumentException if the public publicKey is already registered to a party that does not match the supplied party
      */
```
This kdoc probs needs changing since we've removed the externalId from the method sig

roger-that-dev replied:
ah yes thanks

willhr3 reviewed (changes requested):
Just two minor changes then it looks good to me.

Commit:
Use set instead of first entry in ourNames.
